Custom lattice prefixes

connorsmith256 opened the pull request:
## Feature or Problem
Note: this PR is based on top of https://github.com/wasmCloud/wasmCloud/pull/431

This PR adds support for setting a custom lattice prefix via args

I also moved a couple other settings into `HostConfig`. If we like this approach, I can easily add support for the rest of the host options.

## Related Issues
Resolves https://github.com/wasmCloud/wasmCloud/issues/428

## Release Information
Next

## Consumer Impact
Users can now start the host with a `-x foo-bar` and the `lattice_prefix` for incoming/outgoing messages will match

## Testing
<!---
Declare the testing information for this pull request
--->

<!---
Identify the platforms on which this code was built (include both OS and CPU architecture)
--->
Built on platform(s)
- [ ] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [x] aarch64-darwin
- [ ] x86_64-windows

<!---
Identify the platforms on which this code was tested (include both OS and CPU architecture)
--->
Tested on platform(s)
- [ ] x86_64-linux
- [ ] aarch64-linux
- [ ] x86_64-darwin
- [x] aarch64-darwin
- [ ] x86_64-windows

### Acceptance or Integration
I updated the wasmbus test to use a custom prefix

### Manual Verification
I also tested manually with different prefixes and NATS hosts/ports

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -159,22 +159,40 @@ impl Queue {
             nats.subscribe(format!(
-                "wasmbus.ctl.default.cmd.{}.*",
-                host_key.public_key()
+                "wasmbus.ctl.{prefix}.registries.put",
```
```suggestion
                "wasmbus.ctl.{lattice_prefix}.registries.put",
```

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -159,22 +159,40 @@ impl Queue {
+            nats.subscribe(format!(
+                "wasmbus.ctl.{prefix}.get.{host}.inv",
+                prefix = lattice_prefix,
+                host = host_key.public_key()
```
This should have been called outside the macro

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -334,12 +352,13 @@ struct Provider {
 }
 
-/// Wasmbus lattice
+/// wasmCloud Host
```
Out of scope

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -334,12 +352,13 @@ struct Provider {
     // TODO: Clean up actors after stop
     actors: RwLock<HashMap<String, Arc<Actor>>>,
     cluster_key: KeyPair,
+    config: HostConfig,
```
The config should have been parsed on initialization, only the relevant parts from it should be stored in the resulting struct

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -369,27 +388,23 @@ fn linkdef_hash(
+        config: HostConfig,
     ) -> anyhow::Result<(Arc<Self>, impl Future<Output = anyhow::Result<()>>)> {
-        let cluster_key = if let Some(cluster_seed) = cluster_seed {
+        let cluster_key = if let Some(cluster_seed) = config.cluster_seed.clone() {
```
The `clone` is redundant

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -369,27 +388,23 @@ fn linkdef_hash(
             KeyPair::new(KeyPairType::Cluster)
         };
-        let host_key = if let Some(host_seed) = host_seed {
+        let host_key = if let Some(host_seed) = config.host_seed.clone() {
```
Redundant `clone`

connorsmith256 replied:
Same as above

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -447,8 +464,8 @@ impl Lattice {
         let jetstream = async_nats::jetstream::new(nats.clone());
-        // TODO: Use prefix
-        let bucket = format!("LATTICEDATA_{prefix}", prefix = "default");
+        let lattice_prefix = config.lattice_prefix.clone();
```
Redundant `clone`

connorsmith256 replied:
Fixed

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -1509,7 +1529,7 @@ impl Lattice {
           "prov_rpc_host": "TODO",
           "rpc_host": "TODO",
-          "lattice_prefix": "default",
+          "lattice_prefix": self.config.lattice_prefix,
```
This should have just been a field on the struct

rvolosatovs commented:
Please drop the lattice -> host renaming commit, since it's out-of-scope for this PR and will cause merge conflicts with other work

rvolosatovs commented on tests/wasmbus.rs:
```diff
@@ -49,9 +49,10 @@ async fn wasmbus() -> anyhow::Result<()> {
         )
         .init();
 
+    let nats_host = "localhost";
```
redundant, this is already available from https://docs.rs/url/latest/url/struct.Url.html#method.host

connorsmith256 replied:
I don't think I understand. Are you suggesting we call `Url::parse` first, then extract the host with `.host_str().unwrap()`?

rvolosatovs replied:
Yes, for example. It's confusing to have two "sources of truth", which are likely to diverge over time as we work on this test resulting in very hard-to-debug issues

rvolosatovs replied:
Well, ideally we would not need the `nats_host` at all https://github.com/wasmCloud/wasmCloud/pull/432#discussion_r1289650029

rvolosatovs replied:
We also use it here https://github.com/wasmCloud/wasmCloud/blob/77d663d3e1fd5590177ac8003a313a3edf29ab1f/tests/wasmbus.rs#L191 and https://github.com/wasmCloud/wasmCloud/blob/77d663d3e1fd5590177ac8003a313a3edf29ab1f/tests/wasmbus.rs#L391

If you'd address https://github.com/wasmCloud/wasmCloud/pull/432#discussion_r1289650029, we'd also use it there.

If you would like to define the host name - fine, but please use a constant for that then

rvolosatovs commented on tests/wasmbus.rs:
```diff
@@ -81,18 +82,21 @@ async fn wasmbus() -> anyhow::Result<()> {
     .context("failed to connect to NATS")?;
-    let client = ClientBuilder::new(client)
+
+    let test_prefix = "test-prefix";
```
`const`

connorsmith256 commented:
@rvolosatovs I think I addressed all your review comments. I left a few reply comments for clarification. There's one comment I want to talk about here:

> The config should have been parsed on initialization, only the relevant parts from it should be stored in the resulting struct

I think we should do this, but not right now. Most of what I anticipate will be in `HostConfig` should end up in the `Host` struct, but probably not everything, and I'd rather punt on deciding which things should go where. So for now I'd rather keep those values contained in the `host_config` field, so it will be easy to refactor later

rvolosatovs commented on crates/host/src/wasmbus/config.rs:
```diff
@@ -1,21 +1,27 @@
-    /// URL to connect to
-    pub url: Url,
-    /// Optional host seed
+    /// NATS server host to connect to
```
I think internally, this should not be split into "host" and "port", but be just a single URL.

Apart from simplifying usage internally, a more important aspect of this is the scheme - while I understand that in current OTP implementation it is not possible to force a particular scheme to be used, a developer should be able to e.g. enforce a connection to NATS via TLS, which can only be done by specifying the scheme

connorsmith256 replied:
Can we revisit this later? I mirrored the options we [use in wash](https://github.com/wasmCloud/wash/blob/b0e746be713d070b4400294ec401b87444bd5741/src/up/mod.rs#L94), and this will make it easier to maintain backwards compatibility with tooling that starts hosts. (Whether we use a single URL or a host + port, we'll need to refactor anyway to split that into 3 sets of args: `WASMCLOUD_CTL_*`, `WASMCLOUD_RPC_*`, and `WASMCLOUD_PROV_RPC_*`)

rvolosatovs replied:
Well, we can, but that would mean we have to revert this particular change, since we already are using a URL here, which to me seems as unnecessary churn.

Note, that I'm not asking to change the binary arguments and/or environment variables, just keeping the type of the field on the struct the same as it already was.

rvolosatovs commented on src/config.rs:
```diff
@@ -0,0 +1,17 @@
+/// Default NATS server host
```
Is a separate module really needed for this?

Especially, since you're importing all of these

rvolosatovs replied:
I would rather not try to prematurely optimize this and keep it simple for now

rvolosatovs commented on src/main.rs:
```diff
@@ -4,15 +4,37 @@ use anyhow::{self, Context};
+
+#[allow(clippy::wildcard_imports)]
+use config::*;
+use wasmcloud_host::wasmbus::HostConfig;
```
Could you move this import back together with all the other imports?

rvolosatovs replied:
@connorsmith256 I understand that there is a particular order in other repositories, in ordering in this repository is the following:

- pub use super...
- pub use crate...
- pub use core...
- pub use std...
- pub use everything-else (including other crates in the same workspace)

- use super...
- use crate...
- use core...
- use std...
- use everything-else (including other crates in the same workspace)

Let's please follow the format already used throughout the codebase in this repository and we can always refactor it later. There is no reason to introduce inconsistencies here.

As a part of the refactoring we should introduce a development guideline to avoid such situations in the future

rvolosatovs commented on src/main.rs:
```diff
@@ -4,15 +4,37 @@ use anyhow::{self, Context};
 use tracing_subscriber::prelude::*;
-use wasmcloud_host::WasmbusHostConfig;
+
+mod config;
```
nit; could you move this above the `use` block, that's so far the consistent style used throughout the codebase in this repository.

(It should be made a guideline)

connorsmith256 replied:
Sure

rvolosatovs reviewed (commented):
This is looking much better now, thanks!

rvolosatovs commented on src/main.rs:
```diff
@@ -4,15 +4,37 @@ use anyhow::{self, Context};
+    #[clap(long = "nats-host", default_value = DEFAULT_NATS_HOST, env = ENV_NATS_HOST)]
+    pub nats_host: String,
+    /// NATS server port to connect to
+    #[clap(long = "nats-port", default_value = DEFAULT_NATS_PORT, env = ENV_NATS_PORT)]
```
What do we gain from indirection in all of the `DEFAULT_*` and `ENV*`? Why not define everything directly in the attribute?

connorsmith256 replied:
I followed the pattern we used in wash, and personally, I think it's easier to view the defaults and env vars as their own lists.

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -350,12 +347,13 @@ impl Handler {
         let invocation = self.invocation(operation, request).await?;
         let request =
             rmp_serde::to_vec_named(&invocation).context("failed to encode invocation")?;
+        let lattice_prefix = self.lattice_prefix.clone();
```
```suggestion
        let lattice_prefix = &self.lattice_prefix;
```

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -416,6 +414,7 @@ impl Bus for Handler {
         let (res_r, mut res_w) = socket_pair()?;
 
         let nats = self.nats.clone();
+        let lattice_prefix = self.lattice_prefix.clone();
```
```suggestion
        let lattice_prefix = &self.lattice_prefix;
```

rvolosatovs replied:
This still does not look addressed, but I don't mind fixing that in a follow-up

connorsmith256 replied:
Look below, the usage is in an `async move`. Try to compile this as a ref and you'll get a lifetime error

rvolosatovs replied:
You're right, I've missed that part

rvolosatovs commented on crates/host/src/wasmbus/mod.rs:
```diff
@@ -948,11 +944,13 @@ impl Host {
         });
 
-        let nats = async_nats::connect(url.as_str())
+        let nats_url = format!("{}:{}", config.nats_host, config.nats_port);
```
We already had a rich type here, which was known to be valid, why would we introduce a worse approach here?

rvolosatovs commented on crates/host/src/wasmbus/config.rs:
```diff
@@ -1,21 +1,27 @@
 use serde::{Deserialize, Serialize};
-use url::Url;
 
+#[allow(clippy::module_name_repetitions)]
```
is this really necessary? Could we add a comment here about why?

rvolosatovs commented on tests/wasmbus.rs:
```diff
@@ -226,15 +231,18 @@ async fn wasmbus() -> anyhow::Result<()> {
+        nats_host: nats_host.to_string(),
+        nats_port,
+        lattice_prefix: TEST_PREFIX.to_string(),
+        cluster_seed: cluster_key.seed().ok(),
+        host_seed: host_key.seed().ok(),
```
Can we revert this change? This should panic if the seed key is missing and not silently ignore it - it would introduce a very hard-to-debug issue if any of these keys was not a seed key

rvolosatovs commented on src/main.rs:
```diff
@@ -33,3 +68,15 @@ async fn main() -> anyhow::Result<()> {
+impl From<Args> for WasmbusHostConfig {
+    fn from(args: Args) -> Self {
+        Self {
+            ctl_nats_url: Url::parse(&format!("nats://{}:{}", args.nats_host, args.nats_port))
+                .expect("failed to parse ctl_nats_url"),
```
This should have happened in `main()`, but I don't mind fixing that up in a follow-up myself
